server/service/microsoft_mdm.go

```diff
@@ -0,0 +1,527 @@
+	case mdm_types.GetPoliciesResponse:
+		action = urlPolicy
+		headerXsu = &urlXSU
+		body.Xsi = &urlXSI
+		body.Xsd = &urlXSD
+		body.GetPoliciesResponse = &msg
+
+	case mdm_types.RequestSecurityTokenResponseCollection:
+		action = urlEnroll
+		headerXsu = &urlXSU
+		security = &mdm_types.WsSecurity{
+			MustUnderstand: MUValue,
+			XmlNS:          urlSecExt,
+			Timestamp: mdm_types.Timestamp{
+				ID:      timestampID,
+				Created: getUtcTime(secWindowStartTimeMin), // minutes ago
+				Expires: getUtcTime(secWindowEndTimeMin),   // minutes from now
+			},
+		}
+		body.RequestSecurityTokenResponseCollection = &msg
+
+	case mdm_types.SoapFault:
+		// Setting the target action
+		if msg.OriginalMessageType == mdm_types.MDEDiscovery {
+			action = urlDiscovery
+		} else if msg.OriginalMessageType == mdm_types.MDEPolicy {
+			action = urlPolicy
+		} else if msg.OriginalMessageType == mdm_types.MDEEnrollment {
+			action = urlEnroll
+		} else {
+			action = urlDiag
+		}
+		uuid := uuid.New().String()
+		activityID = &mdm_types.ActivityId{
+			Content:       uuid,
+			CorrelationId: uuid,
+			XmlNS:         urlDiag,
+		}
+		body.SoapFault = &msg
+
+	default:
+		return SoapResponse{}, errors.New("mdm response message not supported")
+	}
+
+	// Return the SoapRequest type with the appropriate fields set
+	return SoapResponse{
+		XmlNSS: urlNSS,
+		XmlNSA: urlNSA,
+		XmlNSU: headerXsu,
+		Header: mdm_types.ResponseHeader{
+			Action: mdm_types.Action{
+				Content:        action,
+				MustUnderstand: MUValue,
+			},
+			RelatesTo:  relatesTo,
+			ActivityId: activityID,
+			Security:   security,
+		},
+		Body: body,
+	}, nil
+}
+
+// mdmMicrosoftDiscoveryEndpoint is the response struct for the GetDiscovery endpoint
+func mdmMicrosoftDiscoveryEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*SoapRequest)
+
+	// Checking first if Discovery message is valid and returning error if this is not the case
+	if err := req.IsValidDiscoveryMsg(); err != nil {
+		return getSoapResponseFault(req.Header.MessageID, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err), nil
+	}
+
+	// Getting the DiscoveryResponse message
+	discoveryMessage, err := svc.GetMDMMicrosoftDiscoveryResponse(ctx)
+	if err != nil {
+		return getSoapResponseFault(req.Header.MessageID, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err), nil
+	}
+
+	// Embedding the DiscoveryResponse message inside of a SoapResponse
+	response, err := NewSoapResponse(discoveryMessage, req.Header.MessageID)
+	if err != nil {
+		return getSoapResponseFault(req.Header.MessageID, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err), nil
+	}
+
+	return response, nil
+}
+
+// GetMDMMicrosoftDiscoveryResponse returns a valid DiscoveryResponse message
+func (svc *Service) GetMDMMicrosoftDiscoveryResponse(ctx context.Context) (*fleet.DiscoverResponse, error) {
+	// skipauth: This endpoint does not use authentication
+	svc.authz.SkipAuthorization(ctx)
```

Comment:
Just for my own understanding, this is where in the future we could (and likely would) validate an access token provided in the opaque JSON as part of the SOAP request, right?

Reply:
yes, that's correct! This is where the access token validation will be performed